Add docs and features to warn about changing indexes in 2.0 (#4830)
cocotb/cocotb: docs/source/update_indexing.rst

.. _update-indexing:

******************************
Update Indexing for cocotb 2.0
******************************

> cocotb 2.0 changed what types were returned when getting a value from a

Reviewer: ```suggestion
cocotb 2.0 changes what types are returned when getting a value from a
```
?

Reviewer: My reasoning for using present tense is that for users having a need to read this section, cocotb 2 is/will be "the new and current thing", regardless of when exactly they read it.

:ref:`Logic-like <logic-handle-value-changes>` or :ref:`Array-like <array-handle-value-changes>` simulator object.
This can require a change in how those values are indexed.
Because this change breaks backwards compatibility silently,
this document and the following set of features were added to assist users in updating the indexing in their existing tests built against cocotb 1.x.

.. autoclass:: cocotb.types.IndexingChangedWarning

.. envvar:: COCOTB_INDEXING_CHANGED_WARNING

    Set this environment variable to ``1`` to cause a warning to be emitted on all :class:`.LogicArray` and :class:`.Array` indexing and slicing operations
    if the indexing would have changed between cocotb 1.x and 2.x.

How to Find Indexing Changes
============================

Set the :envvar:`!COCOTB_INDEXING_CHANGED_WARNING` environment variable to ``1`` and run your test.
You should see warnings emitted on every line of your testbench that indices or slices a :class:`.LogicArray` or :class:`.Array` where the indexing has changed.
The warning will tell you what index you used and what the new index should be.

.. code-block:: console

    $ export COCOTB_INDEXING_CHANGED_WARNING=1
    $ make
    ...

>     /home/user/project/tests/unit_tests.py:160: IndexingChangedWarning: Index 0 is now 7

Author: Do we want it more actionable? `Update index 0 to 3`?

>     /home/user/project/tests/unit_tests.py:160: IndexingChangedWarning: Index 0 is now 7

Reviewer: Yes, I think this would be better.

      left_bit = dut.signal.value[0]
    ...
    /home/user/project/tests/unit_tests.py:543: IndexingChangedWarning: Index 0 is now 3
      if handle.value[i]:
    /home/user/project/tests/unit_tests.py:543: IndexingChangedWarning: Index 1 is now 2
      if handle.value[i]:
    /home/user/project/tests/unit_tests.py:543: IndexingChangedWarning: Index 2 is now 1
      if handle.value[i]:
    /home/user/project/tests/unit_tests.py:543: IndexingChangedWarning: Index 3 is now 0
      if handle.value[i]:
    ...
    /home/user/project/tests/unit_tests.py:232: IndexingChangedWarning: Slice 0:3 is now 7:4
      field = dut.signal.value[:3]

How to Update Indexing
======================

> For static indices or slices, simply change the index or slice to the new value.

Author: Both are correct, we just need to be consistent.

.. code-block:: verilog
    :caption: HDL signal being used

    // Two arrays using the same indices
    logic [3:0] array [1:4];
    logic [4:0] doubled_array [1:4];

.. code-block:: python
    :caption: Looping over indices
    :class: removed

    for i in range(len(dut.array.value)):  # 0, 1, 2, 3
        dut.doubled_array[i].value = 2 * dut.array.value[i]

.. code-block:: python
    :caption: Looping over ``range``
    :class: new

    array_value = dut.array.value
    for i in array_value.range:  # 1, 2, 3, 4
        dut.doubled_array[i].value = 2 * array_value[i]

You can also use the ``array.range`` object to translate ``0`` to ``len()-1`` indexing to the one used by :class:`!LogicArray` and :class:`!Array`.
Prefer this as a last resort, as it is less readable and has overhead compared to the other options.

.. code-block:: python
    :class: new

    val = LogicArray("1010", Range(3, 0))
    assert val[0] == 0      # index 0 is right-most
    ind = val.range[0]      # 0th range value is 3
    assert val[ind] == "1"  # index 3 is left-most

Gradual Update
==============

> The recommended approach to fix this change is to update one module at a time.

Author: Maybe I just leave it out?
```suggestion
The recommended approach is to update one module at a time.
```

> This is done by using the :envvar:`PYTHONWARNINGS` environment variable to limit which modules can emit a :class:`!IndexingChangedWarning`.

Reviewer: ```suggestion
This is done by using the :envvar:`PYTHONWARNINGS` environment variable to limit which modules can emit an :class:`!IndexingChangedWarning`.
```

The :envvar:`!PYTHONWARNINGS` value below will ignore all :class:`!IndexingChangedWarning` except those from the ``unit_tests`` module.

.. code-block:: console

    $ export PYTHONWARNINGS=ignore::cocotb.types.IndexingChangedWarning,default::cocotb.types.IndexingChangedWarning:unit_tests
    $ export COCOTB_INDEXING_CHANGED_WARNING=1
    $ make

After eliminating all warnings from one module, update the :envvar:`!PYTHONWARNINGS` environment variable to the next module and repeat until all modules are updated.

> Then rerun all tests with :envvar:`!PYTHONWARNINGS` unset to ensure all warnings are gone.

Author: I don't think so. Do we support bash, fish, zsh, powershell, etc. etc. or just trust the user can Google it?
